### PR TITLE
CORE-9052 beta1 to 5.0 2022.12.22 merge

### PR DESCRIPTION
**Pull request by vlajos**
CORE-9052
* Resolve input and reference input states in UtxoLedgerTransactionFactory.
* Which requires the UtxoLedgerPersistenceService. To avoid creating a circular dependency between SignedTxFactory and LedgerPersistenceService, we made the persistence service accessible via non-osgi. This needs to improved later.
* Removed some toLedgerTransaction() calls where the already available signed transaction has all the required information in the given context. (To spare the database lookups and adding @Suspendable)
* WrappedUtxoWireTransactionKryoSerializer
* ReferenceInputState -> ReferenceState

CORE-9014 
* Fixed issue with incorrect flow external event ID not set correctly for token claim release acknowledgements
* Fixed issue preventing flow resuming after failed token query.
* Fixed issue preventing token selection API flow injection.
* Removed old unused test code

CORE-9021
* Enable notary smoke test scenarios

CORE-9050
* Fix bug with Kryo serialization

API: https://github.com/corda/corda-api/pull/763

### CODE REVIEW

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/persistence/UtxoLedgerPersistenceServiceImpl.kt [hunk 1]
```diff
@@ -128,8 +128,9 @@ class UtxoLedgerPersistenceServiceImpl @Activate constructor(
         }
     }
 
-    private fun SignedTransactionContainer.toSignedTransaction(): UtxoSignedTransaction {
-        return utxoSignedTransactionFactory.create(wireTransaction, signatures)
+    private fun SignedTransactionContainer.toSignedTransaction(utxoLedgerPersistenceService: UtxoLedgerPersistenceService)
+    : UtxoSignedTransaction {
+        return utxoSignedTransactionFactory.create(wireTransaction, signatures, utxoLedgerPersistenceService)
```
Comment (relyafi) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/persistence/UtxoLedgerPersistenceServiceImpl.kt]:
```suggestion
        return utxoSignedTransactionFactory.create(wireTransaction, signatures, this)
```

Seems a little odd to take `utxoLedgerPersistenceService` as an argument, when it is `this` object and this function is private? Suggested approach results in being able to revert the change to line 62 as well.

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoSignedTransactionImpl.kt [hunk 2]
```diff
@@ -103,15 +104,15 @@ data class UtxoSignedTransactionImpl(
...
             }
         }.map { it.by }.toSet()
-        val requiredSignatories = this.toLedgerTransaction().signatories
-        if (requiredSignatories.any {
+        if (signatories.any {
                 !it.isFulfilledBy(appliedSignatories) // isFulfilledBy() helps to make this working with CompositeKeys.
-            }){
+            }) {
             throw TransactionVerificationException(id, "There are missing signatures", null)
         }
```
Comment (relyafi) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoSignedTransactionImpl.kt]: Something looks off with the indentation here (think this was also the case before. Would suggest moving the opening brace to the next line., or adding another layer of indentation for the throw statement.

Also, the error may be more useful if it said which expected signatures are missing?

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoTransactionBuilderImpl.kt [hunk 3]
```diff
@@ -161,7 +163,7 @@ data class UtxoTransactionBuilderImpl(
                 && other.attachments == attachments
                 && other.commands == commands
                 && other.inputStateRefs == inputStateRefs
-                && other.referenceInputStateRefs == referenceInputStateRefs
+                && other.referenceStateRefs == referenceStateRefs
```
Comment (relyafi) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoTransactionBuilderImpl.kt:166]: Not new to this change - but does this imply that for two of these to be equal, the exact order of these must be the same for two objects to be equal? I.e. if I had two objects that were the same other than one defined input states as [A,B] and another defined them as [B,A], should these be considered equal, or different?

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/factory/UtxoLedgerTransactionFactory.kt [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package net.corda.ledger.utxo.flow.impl.transaction.factory
+
+import net.corda.ledger.common.data.transaction.WireTransaction
+import net.corda.v5.base.annotations.Suspendable
+import net.corda.v5.ledger.utxo.transaction.UtxoLedgerTransaction
+
+interface UtxoLedgerTransactionFactory {
```
Comment (relyafi) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/factory/UtxoLedgerTransactionFactory.kt:12]: Think this needs a kdoc comment.

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/factory/impl/UtxoLedgerTransactionFactoryImpl.kt [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+            wrappedUtxoWireTransaction.inputStateRefs.map { it ->
+                sourceTransactions[it.transactionHash]?.get(it.index)
+                    ?: throw (CordaRuntimeException("Input state not found ${it.transactionHash} ${it.index}"))
+            }
+
+        val referenceStateAndRefs =
+            wrappedUtxoWireTransaction.referenceStateRefs.map { it ->
+                sourceTransactions[it.transactionHash]?.get(it.index)
+                    ?: throw (CordaRuntimeException("Reference state not found ${it.transactionHash} ${it.index}"))
+            }
```
Comment (relyafi) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/factory/impl/UtxoLedgerTransactionFactoryImpl.kt]: This is probably the most elegant solution, but also looks like it's doing an unnecessary additional iteration of the input and reference states. Might be more efficient to do one loop each of these and to setup the source transactions and relevant stateandrefs object in one loop.

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/factory/impl/UtxoLedgerTransactionFactoryImpl.kt [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+                .map { it.transactionHash }
+                .toSet()
+                .associateWith { it ->
+                    utxoLedgerPersistenceService.find(it)?.outputStateAndRefs
+                        ?: throw (CordaRuntimeException("Could not find transaction $it when fetching input states."))
+                }
+        val inputStateAndRefs =
+            wrappedUtxoWireTransaction.inputStateRefs.map { it ->
+                sourceTransactions[it.transactionHash]?.get(it.index)
+                    ?: throw (CordaRuntimeException("Input state not found ${it.transactionHash} ${it.index}"))
```
Comment (relyafi) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/factory/impl/UtxoLedgerTransactionFactoryImpl.kt]:
```suggestion
                    ?: throw (CordaRuntimeException("Input state not found ${it.ref}"))
```

Should use the standard `toString()` defined on the reference part, which will add a colon between the hash and index and ensures formatting consistency.

Comment also applies to line 51 as well.

**Review by relyafi — COMMENTED** [review]
Notary bits LGTM, have some general comments.

Appreciate this is a merge PR so might need some thought as to how and where to address these comments if you decide to fix.

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/factory/impl/UtxoLedgerTransactionFactoryImpl.kt (hunk 5 above)
Comment (blsemo) [reply components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/factory/impl/UtxoLedgerTransactionFactoryImpl.kt]: I am not so sure about this one - the expensive bit here is finding the source transaction, not looping through the inputs/references. 
So either we need one iteration caching all txs temporarily, or this two phase approach. A comment explaining the reasoning might be useful, though.

**Comment by vlajos** [comment]
> 



> Appreciate this is a merge PR so might need some thought as to how and where to address these comments if you decide to fix.

Thanks. I am addressing them here: https://github.com/corda/corda-runtime-os/pull/2819

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/persistence/UtxoLedgerPersistenceServiceImpl.kt (hunk 1 above)
Comment (vlajos) [reply components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/persistence/UtxoLedgerPersistenceServiceImpl.kt]: Updated here: https://github.com/corda/corda-runtime-os/pull/2819/files#diff-6ca8a2c0e6178df692e3bc6b4d24858f329a12a609ffa359752d20bc07971377

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoSignedTransactionImpl.kt (hunk 2 above)
Comment (vlajos) [reply components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoSignedTransactionImpl.kt]: Updated here: https://github.com/corda/corda-runtime-os/pull/2819/files#diff-ba66f90f1d378b1fc768f02d19f5ec2b6ed7c1c6b947c3e0cc5710f8a15c4afb
The logging side has already a ticket: https://r3-cev.atlassian.net/browse/CORE-8935

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoTransactionBuilderImpl.kt (hunk 3 above)
Comment (vlajos) [reply components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoTransactionBuilderImpl.kt:166]: AFAIK if the order of inputs is different the two transactions are supposed to be treated as different ones.
Their merkle trees and txids will be different without sorting them anyway.

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/factory/UtxoLedgerTransactionFactory.kt (hunk 4 above)
Comment (vlajos) [reply components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/factory/UtxoLedgerTransactionFactory.kt:12]: I have added few lines here: https://github.com/corda/corda-runtime-os/pull/2819/files#diff-4ee14d284a0c0ef2c15f2c0760c832534c7662f5cef6c4a5eeac36f786959096
(It is not in the public API.)

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/factory/impl/UtxoLedgerTransactionFactoryImpl.kt (hunk 6 above)
Comment (vlajos) [reply components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/factory/impl/UtxoLedgerTransactionFactoryImpl.kt]: Thanks. I have updated them:
https://github.com/corda/corda-runtime-os/pull/2819/files#diff-e512a249344bb1ddd2949e337a93c0e25ef4c25b2af08a23a643bc1715c04122

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoTransactionBuilderImpl.kt (hunk 3 above)
Comment (blsemo) [reply components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoTransactionBuilderImpl.kt:166]: Technically, the order doesn't matter - but I don't think it's a big issue. I would normally only expect transaction builders to be equal if they are copies of the same builder, in which case the order is the same. 
I can't really see a use case where someone would create two equal transation builders completely independently.
One might even argue that the order of states in these lists matters, as re-ordering them changes the Merkle tree of the SignedTransaction that will be created.

Comment (relyafi) [reply components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoTransactionBuilderImpl.kt:166]: I don't have a strong opinion either way, just wanted to ensure we'd considered what our expected / correct behavior is.

All I'd say is that we never know how customers will use our objects, so we should be prepared to justify our behavior either way.

**Review by relyafi — DISMISSED** [review]

**Review by lankydan — APPROVED** [review]
